cos_agent interface (#54)
canonical/charm-relation-interfaces: interfaces/cos_agent/v0/README.md

# `cos_agent`

## Usage

This relation interface describes the expected behavior of any charm claiming to be able to provide or require the `cos_agent` interface.

In most cases, this will be accomplished using the [`cos_agent` library](https://charmhub.io/grafana-agent/libraries/cos_agent), although charm developers are free to provide alternative libraries as long as they fulfill the behavioral and schematic requirements described in this document.

## Direction

```mermaid
flowchart TD
    Provider -- "relation_name
    metrics_endpoints
    metrics_rules_dir
    logs_rules_dir
    recursive_rules_dir
    log_slots
    dashboard_dirs
    refresh_events" --> Requirer
    Requirer -- refresh_events --> Provider
```

As all Juju relations, the `cos_agent` interface consists of a provider and a requirer. In this case the `Provider` side of the relationship may provide telemetry settings, if not, the required side will use default values.

## Behavior

> Since this interface is meant to be simple and lightweight each telemetry parameter in the `Provider` side is optional, if some of these are not provided, dafault values will be used.

Reviewer: **thought:** Iiuc, we do not need to mention defaults here because they are not relevant to the spec.

Defaults are a particular lib's concerns, not the schema.

### Provider

- Is expected to be able to provide metrics endpoints to be scraped.
- Is expected to be able to provide a directory containing Prometheus style metrics rules files.
- Is expected to be able to provide a directory containing Loki style rules files.
- Is expected to be able to inform if metrics and logs rules directories should be scanned recursively.
- Is expected to be able to provide snap slots to connect to for scraping logs.
- Is expected to be able to provide a list of directories containing Grafana dashboard files.

> - Is expected to be able to provide a list of events on which to refresh relation data.

Reviewer: **question:** Are those library specs or interface specs?

Author: I think are both...

Reviewer: Shouldn't this be strictly interface specs?
I.e. the contents of the `*.json` has nothing to do with
- a directory containing Prometheus style metrics rules files
- inform if metrics and logs rules directories should be scanned recursively
- etc.

but I guess I'm missing something..?

### Requirer
- Is expected to be able to scrape Prometheus metrics from metrics endpoints and remote-write these metrics to Prometheus and Prometheus compatible systems.
- Is expected to be able to forward alert rules exposed over the relation data bag to Prometheus.
- Is expected to be able to forward logs from the Provider to Loki.
- Is expected to be able to forward alert rules exposed over the relation data bag to Loki.
- Is expected to be able to forward Grafana Dashboards exposed over the relation data bag to Grafana.

## Relation Data

### Provider

[\[JSON Schema\]](./schemas/provider.json)

- Exposes all scrape jobs the requirer should scrape metrics through. Should be placed in the **application** databag.

> - Exposes the unit address of each unit to scrape, as well as the unit name of each address. Should be placed in the **unit** databag of each scrapable unit.

Reviewer: **suggestion:** Split this into two sections: application, unit.

#### Example

> ```yaml

Reviewer: **typo:** Looks like json.

application-data:
  alert_rules: {
    "groups": [
      {
        "name": "an_alert_rule_group",
        "rules": [
          {
            "alert": "SomethingIsUp",
            "expr": "something_bad == 1",
            "for": "0m",
            "labels": {
              "some-label": "some-value"
            },
            "annotations": {
              "some-annotation": "some-other-value"
            }
          }
        ]
      }
    ]
  }
  scrape_jobs: [
    {
      "metrics_path": "/metrics",
      "static_configs": [
        { "targets": ["*:4080"] }
      ]
    }
  ]
  scrape_metadata: {
    "model": "cos",
    "model_uuid": "c2e9f4d5-dcb3-4870-8509-330eb9745ee8",
    "application": "zinc-k8s",
    "unit": "zinc-k8s/0",
    "charm_name": "zinc-k8s"
  }
related-units:
  zinc-k8s/0:
    data:
      prometheus_scrape_unit_address: zinc-k8s-0.zinc-k8s-endpoints.cos.svc.cluster.local
      prometheus_scrape_unit_name: zinc-k8s/0
      # ...
    # ...
  zinc-k8s/1:
    data:
      prometheus_scrape_unit_address: zinc-k8s-1.zinc-k8s-endpoints.cos.svc.cluster.local
      prometheus_scrape_unit_name: zinc-k8s/1
      # ...
    # ...
```

### Requirer

> No relation data should be exposed by the requirer of this relation.

Reviewer: **suggestion (non-blocking):**
```suggestion
None.
```

# `cos_agent`

## Usage

> The `cos_agent` interface loads observability-related file paths and metrics endpoints into relation data.

Reviewer 2: ```suggestion
The `cos_agent` interface is meant to exchange information on several observability-related data streams at once.
```

> The first implementation was for the [`grafana-agent`](https://charmhub.io/grafana-agent/libraries/cos_agent) machine charm for "converting" paths to log files, alert rules and dashboards into relation data for the `loki_push_api`, `prometheus_remote_write` and `grafana_dashboards` interfaces.

Reviewer 2: While I appreciate the historical note, this is not part of the specification. I'm wondering if we should however make it clear that this is a 'private' interface, not quite meant to be generic and usable by other charms to talk to some observability charms. 
@simskij WDYT?

Reviewer 2: Also we should make it more clear that this is a SUBORDINATE, ergo MACHINE-charm-specific relation. Should these have a special place in charm-relation-interfaces?
I'm actually thinking we might need some special tags and metadata to distinguish them from regular relations, or the interface tester might behave incorrectly. But that's on me, once we implement tests.

Reviewer 3: Yes, I think we should. Let's list it in a separate table in the readme, just like the identity and mlops solution-specific interfaces.

Author: Done

```mermaid
flowchart TD
    Provider["Provider (e.g. kafka)"] -- "relation_name
    metrics_endpoints
    metrics_rules_dir
    logs_rules_dir
    recursive_rules_dir
    log_slots
    dashboard_dirs
    refresh_events" --> Requirer
    Requirer["Requirer (e.g. grafana-agent)"] -- refresh_events --> Provider
```

As all Juju relations, the `cos_agent` interface consists of a provider and a requirer. In this case the `Provider` side of the relationship may provide telemetry settings, if not, the required side will use default values.

## Behavior

### Provider

- Is expected to be able to provide metrics endpoints to be scraped.
- Is expected to be able to provide a directory containing Prometheus style metrics rules files.
- Is expected to be able to provide a directory containing Loki style rules files.
- Is expected to be able to inform if metrics and logs rules directories should be scanned recursively.
- Is expected to be able to provide snap slots to connect to for scraping logs.
- Is expected to be able to provide a list of directories containing Grafana dashboard files.
- Is expected to be able to provide a list of events on which to refresh relation data.

### Requirer

> - Is expected to be able to scrape Prometheus metrics from metrics endpoints and remote-write these metrics to Prometheus and Prometheus compatible systems.

Reviewer 2: I'm wondering if here we should be focusing on the behaviour rather than on the actor.
I.e. instead of "is expected to forward this data to Loki": "is expected to [do whatever loki would do]". The fact that we 'subcontract' the logic to another charm is not relevant. The requirer could in principle do all work itself and we wouldn't notice nor care.

Author: You have a solid point here. We should focus on the behaviour rather than the actor.
With that said, we can remove completely this section.

- Is expected to be able to forward alert rules exposed over the relation data bag to Prometheus.
- Is expected to be able to forward logs from the Provider to Loki.
- Is expected to be able to forward alert rules exposed over the relation data bag to Loki.
- Is expected to be able to forward Grafana Dashboards exposed over the relation data bag to Grafana.

## Relation Data

### Provider

[\[JSON Schema\]](./schemas/provider.json)

#### Application data

> - Exposes all scrape jobs the requirer should scrape metrics through. Should be placed in the **application** databag.

Reviewer 2: ```suggestion
- Exposes all scrape jobs the requirer should scrape metrics through.
```

We're already in the 'application data' section

#### Unit data

> - Exposes the unit address of each unit to scrape, as well as the unit name of each address. Should be placed in the **unit** databag of each scrapable unit.

Reviewer 2: ```suggestion
- Exposes the unit address of each unit to scrape, as well as the unit name of each address. 
```
idem

#### Example

```json

> application-data:

Reviewer 2: this is not valid json
